### DISCUSSION TITLE
Normalization and Cost

### DISCUSSION THREAD

**Discussion by horsto** [deepcharles/ruptures#312] [Q&A]
Hi, this is a question, not an issue.
I have a bunch of features that I track over time. I am feeding them into 
```python
algo = rpt.Pelt(model=model, min_size=1, jump=1)
algo.fit(signal)
result = algo.predict(pen=p) # RESULT OF CHANGE POINT DETECTION
```
`signal` here is (for example) a 500x16 (timepoints x features). The features themselves live on pretty different scales, such that I thought that some kind of scaling / normalization (for example via https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.scale.html#sklearn.preprocessing.scale) could make sense. Now I wonder though how different costs would be affected by that. In the example I am attaching below you can see the normalized signal for L1 and L2 norms -> change points are depicted with dashed lines. You can see that there are some obvious misses there (calibrating the penalty helps sometimes, but is a finicky process). 
Should normalization be skipped altogether / is there a better alternative cost for these kind of signals? 

<img width="1254" alt="Screenshot 2023-10-11 at 15 36 46" src="https://github.com/deepcharles/ruptures/assets/7548928/850bb788-ff0e-439d-9bd7-8d28c7a9f92c">

**Comment by tg12** [comment]
What are you using to draw these graphs as an unrelated question!

> Should normalization be skipped altogether / is there a better alternative cost for these kind of signals?

 I do agree in some instances there might be a need to remove any pre processing of the data, this can be done upstream if needed unless it's an inherent part of the pelt algorithm.

**Comment by horsto** [comment]
It's not inherent to the pelt algorithm I think? Unless there is some hidden pre processing going on (?). 

I would like to know whether I should do my own normalization up front, and how it might affect certain cost functions in the pelt algorithm (L1, L2, ...). 

The plotting is just matplotlib + seaborn!

**Comment by deepcharles** [comment]
Hi,

Sorry for the late reply.

To normalize or not is task-dependant and there is no definite answer. For multivariate signals, PELT will detect the largest shifts, i.e., those with a large norm `||m_before - m_after||` where `m_before` and `m_after` are the multivariate averages just before and after the change. 
As an example, consider the following 2D signal.
![raw](https://github.com/deepcharles/ruptures/assets/14997422/90c9c3dd-23b3-4e9f-8101-99a7068ec3a7)

One dimension has large shifts and the other has small shifts. Without normalization, only changes in the large dimension are detected.

```python
rpt.display(s, [], rpt.Pelt().fit(s).predict(pen=50))
```

![bkps1](https://github.com/deepcharles/ruptures/assets/14997422/8cac49bb-5124-4b43-93f4-26afe24710fb)


After normalization, all changes are detected.
```python
rpt.display(s, [], rpt.Pelt().fit(s).predict(pen=50))
```
![bkps2](https://github.com/deepcharles/ruptures/assets/14997422/acfa5aa0-942c-47ad-aea5-7eb114ba6779)

Hope this helps

**Comment by horsto** [comment]
Thanks @deepcharles, that makes sense. 
I have added some more screenshots for how `Pelt` behaves for L1 and L2 costs on my signal. Penalty is 80 for both, lower penalties seem to add more noise. There are some misses (as in the first post above).  
There is no difference between normalizing mean, but scaling the data to unit variance does make a difference. This is done here with [sklearn.preprocessing.scale](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.scale.html#sklearn.preprocessing.scale).

<img width="1312" alt="Screenshot 2023-10-24 at 08 37 49" src="https://github.com/deepcharles/ruptures/assets/7548928/75ffbffc-576e-4670-8b9c-ffa411e72bb1">
<img width="1309" alt="Screenshot 2023-10-24 at 08 38 57" src="https://github.com/deepcharles/ruptures/assets/7548928/24ae615c-3365-46e9-836b-108a078e761b">

No normalization above shows the raw signal. 

I guess fine tuning the penalty will do the trick. Or is there anything that I am missing (a better cost function for my signal for example)? Thanks for your help!
